### PR TITLE
Add manual specification of file type to data::Load().

### PR DESCRIPTION
**Pull request by rcurtin**
This solves #2601, #2596, and #2084.

It's originally based on work in #2135; thanks @Shikhar-S for starting that work!  If you like, we can add your name to the codebase since you did part of the work. :+1:

Previously, the `data::Load()` and `data::Save()` functions detected the file type entirely based on the extension, partially using internal Armadillo functionality to do so.  This PR adds an extra optional parameter to `data::Load()` and `data::Save()` of type `arma::file_type`, that can be used to override the automatic detection.  If automatic detection is desired, then `arma::auto_detect` should be used.

Given that mlpack will now break with the newest Armadillo, this, along with #2600, should be merged and then a bugfix release should be done.

### CODE REVIEW

File: HISTORY.md [hunk 1]
```diff
@@ -133,6 +137,7 @@
   * The DecisionStump class has been marked deprecated; use the `DecisionTree`
     class with `NoRecursion=true` or use `ID3DecisionStump` instead (#2099).
 
+>>>>>>> origin/master
```
Comment (coatless) [inline HISTORY.md:140]:
```suggestion
```

Comment (rcurtin) [reply HISTORY.md:140]: What?  How did this get here.  I thought I removed it?  Anyway, nice catch. :)

**Comment by Shikhar-S** [comment]
> It's originally based on work in #2135; thanks @Shikhar-S for starting that work! If you like, we can add your name to the codebase since you did part of the work. 👍

Thanks! Please go ahead.

File: src/mlpack/core/data/detect_file_type.cpp [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+  f.clear();
+  f.seekg(pos1);
+
+  // Handle empty files.
+  if (nMax == 0)
+    return arma::file_type_unknown;
+
+  const arma::uword nUse = std::min(nMax, arma::uword(4096));
+
+  unsigned char* dataMem = new unsigned char[nUse];
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/core/data/detect_file_type.cpp:77]: That's true, but personally I think it's fine now.  When I resurrected #2135, I didn't closely check the memory usage, but the static code analysis job got it, and now when I look through I don't see any further issues. :+1:

File: src/mlpack/core/data/detect_file_type.cpp [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  }
+  else if (extension == "txt")
+  {
+    // This could be raw ASCII or Armadillo ASCII (ASCII with size header).
+    // We'll let Armadillo do its guessing (although we have to check if it is
+    // arma_ascii ourselves) and see what we come up with.
+
+    // This is taken from load_auto_detect() in diskio_meat.hpp
+    const std::string ARMA_MAT_TXT = "ARMA_MAT_TXT";
+    // char* rawHeader = new char[ARMA_MAT_TXT.length() + 1];
```
Comment (zoq) [inline src/mlpack/core/data/detect_file_type.cpp:192]: Do you like to keep the line for reference?

File: src/mlpack/core/data/detect_file_type.cpp [hunk 4]
```diff
@@ -0,0 +1,290 @@
+/**
+ * @file detect_file_type.cpp
```
Comment (zoq) [inline src/mlpack/core/data/detect_file_type.cpp:2]:
```suggestion
 * @file core/data/detect_file_type.cpp
```

File: src/mlpack/core/data/detect_file_type.hpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
+/**
+ * @file detect_file_type.hpp
```
Comment (zoq) [inline src/mlpack/core/data/detect_file_type.hpp:2]:
```suggestion
 * @file core/data/detect_file_type.hpp
```

File: src/mlpack/core/data/detect_file_type.cpp (hunk 3 above)
Comment (rcurtin) [reply src/mlpack/core/data/detect_file_type.cpp]: Good point, removed in 065ae8b9.

**Review by zoq — APPROVED** [review]
Looks great to me, thanks.

**Comment by rcurtin** [comment]
The memory checks failure appears to be because polar went offline during the build, so I won't worry about it. :+1:

**Pull request merged**